Add support for parallel gitian building

superbaud opened the pull request:
To try it out, run `git clone https://github.com/superbaud/zcash-gitian -b parallel-build` (or do `git switch parallel-build` in your `zcash-gitian` repo), and set the `.env` therein to:
```
GPG_KEY_ID=<your-key-id>
GPG_KEY_NAME=<your-key-name>
ZCASH_GIT_REPO_URL=https://github.com/superbaud/zcash
ZCASH_VERSION=parallel-gitian-descriptor

GITIAN_BUILDER_URL=https://github.com/devrandom/gitian-builder
GITIAN_BUILDER_VERSION=master
``` 

Until the corresponding PR (https://github.com/zcash/zcash/pull/6104) for the zcash repository is merged, you will need to target a `ZCASH_VERSION` which has the new parallel gitian descriptor included.

Run `vagrant destroy zcash-build && vagrant up zcash-build`, and once the Vagrant VM is up, run `./gitian-parallel-build.sh`. `./gitian-build.sh` works as previously.

To be sure that we still have reproducibility, it would be good if someone else ran both `./gitian-build.sh`  and `./gitian-parallel-build.sh` (**in clean instances of the Vagrant VM, be sure you run** `vagrant destroy zcash-build` **before each time you run either command**) targeting the above repo URL and branch name to verify that they get the same hashes as I do.

## Edit the `Vagrantfile` so the `v.memory` and `v.cpus` values are a bit lower than the memory and CPU count in the actual physical machine, if they're too high then you might wedge your system badly.

charlieok commented:
I had some unsuccessful runs last week but I tried this again today with the settings in the description here and got a run to completion with the following hashes:

buster:
```
37a483166c143663cdb03cdd0eb82e1ff9ea755a39aacb19c83ce21c5188b5b7  src/zcash-5.2.0.tar.gz
ef1fe3d30a20097a94c0ba242e59262f8fd0dad9336c5c8972c99520e42be762  zcash-5.2.0-linux64-debug.tar.gz
5485e4f2924fdef4afbeba1b30dcf521e84614107d48242695c68c226338eb55  zcash-5.2.0-linux64.tar.gz
```

bullseye:
```
74ecf49e8edbe694216ecca2b2623df7520a59cdd26c68bf2bc1bb051630cced  src/zcash-5.2.0.tar.gz
e71f25288c2fd9e548732bd2e2235f9029ff014cfbcfb4d35a7e1c63446f89e7  zcash-5.2.0-linux64-debug.tar.gz
ff4174bcdb7068756c33e8103293ddc76e0221a9df50124bf0726975c9e56e1e  zcash-5.2.0-linux64.tar.gz
```

superbaud commented:
Hm, what were the errors? Could it have been the `apt-cacher-ng` issue fixed in the other PR?

charlieok commented:
I think it was issues in the early 'getting dependencies' stage so that's very possible.

superbaud commented:
Machine 1 output:

```
vagrant@zcash-build:~$ head gitian.sigs/parallel-gitian-descriptor_bu*/sasha/*assert
==> gitian.sigs/parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> gitian.sigs/parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
vagrant@zcash-build:~$ free -m
               total        used        free      shared  buff/cache   available
Mem:            1982         106          68           0        1807        1731
Swap:           1023          34         989
vagrant@zcash-build:~$ 
```

superbaud commented:
Machine 2 output:
```
vagrant@zcash-build:~/gitian.sigs$ head parallel-gitian-descriptor_bu*/*/*assert
==> parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
vagrant@zcash-build:~/gitian.sigs$ free -m
               total        used        free      shared  buff/cache   available
Mem:           16007         147        5654           0       10206       15532
Swap:           1023           0        1023
vagrant@zcash-build:~/gitian.sigs$ 
```

superbaud commented:
I cleaned up the commits and updated the documentation in `README.md`. I don't think my changes should have changed the functionality of this PR but in order to be confident this is the case, I am currently running a test of ac0bf7d32008cbf00059477062e24cc59c4686a8 on two different machines, with `.env` looking like this (`parallel-gitian-descriptor` points to commit cdac686da4d559af156e3ebb757f8ff981a1f7bc)
```
GPG_KEY_ID=16460E1BC0E0507E
GPG_KEY_NAME=sasha
ZCASH_GIT_REPO_URL=https://github.com/superbaud/zcash
ZCASH_VERSION=parallel-gitian-descriptor

GITIAN_BUILDER_URL=https://github.com/devrandom/gitian-builder
GITIAN_BUILDER_VERSION=master
```

Please let me know if there are questions, reviews (blocking or otherwise), or concerns with this PR, or the accompanying zcash PR (https://github.com/zcash/zcash/pull/6104)!

superbaud commented:
From my amd64 laptop:
```
vagrant@zcash-build:~$ free -m
               total        used        free      shared  buff/cache   available
Mem:            3931          91         116           0        3722        3601
Swap:           1023           0        1023
vagrant@zcash-build:~$ !hea
head gitian.sigs/parallel-gitian-descriptor_bu*/sasha/*assert
==> gitian.sigs/parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> gitian.sigs/parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
vagrant@zcash-build:~$ 
```

superbaud commented:
from GCP machine:
```
vagrant@zcash-build:~$ head gitian.sigs/parallel-gitian-descriptor_bu*/sasha/*assert;free -m
==> gitian.sigs/parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> gitian.sigs/parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
               total        used        free      shared  buff/cache   available
Mem:            7955         136        1181           0        6638        7522
Swap:           1023           1        1022
vagrant@zcash-build:~$ 
```

superbaud commented:
GCP machine
```
vagrant@zcash-build:~/gitian.sigs$ free -m; head parallel-gitian-descriptor_bu*/sasha/*assert
               total        used        free      shared  buff/cache   available
Mem:           16007         163        5506           0       10337       15515
Swap:           1023           0        1023
==> parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
vagrant@zcash-build:~/gitian.sigs$ 
```

superbaud commented:
laptop:
```
vagrant@zcash-build:~/gitian.sigs$ free -m; head parallel-gitian-descriptor_bu*/sasha/*assert
               total        used        free      shared  buff/cache   available
Mem:            3931         101         124           0        3705        3590
Swap:           1023           0        1023
==> parallel-gitian-descriptor_bullseye/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    b64a7e50c203c296df84dc2220c8bcbde14816bd6fd7138dce435efcb6591a92  src/zcash-5.3.0-rc1.tar.gz
    ceed0ae4ebc2221a1e378d07092789589839fe3a2f9ad17d649fdf50e0a21c15  zcash-5.3.0-rc1-linux64-debug.tar.gz
    6a92c9112afab3009731631bc6968f4c1d635766e5cca85b72a33205d2a31fdc  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    39d73b02c143950c9e0e7329e1fa5268ff2534380eea85d3c72348a9994661d5  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - bullseye-amd64: |

==> parallel-gitian-descriptor_buster/sasha/zcash-5.3.0-rc1-build.assert <==
--- !!omap
- out_manifest: |
    dddb25a79fe3558d0203f83660a79e0ee2072de95a45c6156134b74a4a789261  src/zcash-5.3.0-rc1.tar.gz
    a8d6c10e3b6e02e817b518d32ae73032c4e5debbc2140a56430f81a68f50fb9b  zcash-5.3.0-rc1-linux64-debug.tar.gz
    533d3393f60b061cbdbcbb5a699b93dd7755687bb6fbdb6a994e03f7bee6056b  zcash-5.3.0-rc1-linux64.tar.gz
- in_manifest: |-
    4933a908ac8cf0588cf53af775a85abd0c659dd78c8a69754ea5735268bcdb17  zcash-5.3.0-rc1-desc.yml
    git:cdac686da4d559af156e3ebb757f8ff981a1f7bc zcash
- base_manifests: !!omap
  - buster-amd64: |
vagrant@zcash-build:~/gitian.sigs$ 
```
